src/julia-syntax.scm

```diff
@@ -2075,26 +2087,43 @@
-                (let* ((xx  (if (or (and (symbol? x) (not (memq x lhss)))
-                                    (ssavalue? x))
-                                x (make-ssavalue)))
-                       (ini (if (eq? x xx) '() (list (sink-assignment xx (expand-forms x)))))
-                       (n   (length lhss))
-                       (st  (gensy)))
-                  `(block
-                    (local ,st)
-                    ,@ini
-                    ,.(map (lambda (i lhs)
-                             (expand-forms
-                              (lower-tuple-assignment
-                               (if (= i (- n 1))
-                                   (list lhs)
-                                   (list lhs st))
-                               `(call (top indexed_iterate)
-                                      ,xx ,(+ i 1) ,.(if (eq? i 0) '() `(,st))))))
-                           (iota n)
-                           lhss)
-                    (unnecessary ,xx))))))
+                (begin
+                  ;; like memq, but if last element of lhss is (... sym),
+                  ;; check against sym instead
+                  (define (in-lhs? x lhss)
+                    (if (null? lhss)
+                        #f
+                        (let ((l (car lhss)))
+                          (cond ((and (pair? l) (eq? (car l) '|...|))
+                                 (if (null? (cdr lhss))
+                                     (eq? (cadr l) x)
+                                     (error (string "invalid \"...\" on non-final assignment location \""
+                                                    (cadr l) "\""))))
+                                ((eq? l x) #t)
+                                (else (in-lhs? x (cdr lhss)))))))
+                  ;; in-lhs? also checks for invalid syntax, so always call it first
+                  (let* ((xx  (if (or (and (not (in-lhs? x lhss)) (symbol? x))
+                                      (ssavalue? x))
+                                  x (make-ssavalue)))
+                         (ini (if (eq? x xx) '() (list (sink-assignment xx (expand-forms x)))))
+                         (n   (length lhss))
+                         (st  (gensy)))
+                    `(block
+                      (local ,st)
+                      ,@ini
+                      ,.(map (lambda (i lhs)
+                               (expand-forms
+                                 (if (and (pair? lhs) (eq? (car lhs) '|...|))
+                                     `(= ,(cadr lhs) (call (top rest) ,xx ,.(if (eq? i 0) '() `(,st))))
+                                     (lower-tuple-assignment
+                                      (if (= i (- n 1))
+                                          (list lhs)
+                                          (list lhs st))
+                                      `(call (top indexed_iterate)
+                                             ,xx ,(+ i 1) ,.(if (eq? i 0) '() `(,st)))))))
```

Comment:
This stumped me quite a bit. Shouldn't `` `(,st) `` be exactly the same as `st`? Somehow Julia breaks if I replace this though.

Reply:
``` `(,st)``` is the same as `(list st)`, and `,.` splats lists. It's the mutating version of `,@`, and so a kind of old-fashioned micro optimization. `,.` should probably not be used anymore.

Reply:
Ah, that makes total sense! I think I was just confusing it with the way Julia does unquoting, but of course in Lisp, parentheses are actually relevant. Thanks!